### PR TITLE
[Resubmit] Remove mmap/mremap/munmap from Allocator.h

### PR DESCRIPTION
**Pull request by nickitat**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Remove manual calls to `mmap/mremap/munmap` and delegate all this work to `jemalloc`.

---

Resubmit of #51686 and all the credit for the actual optimisation goes to @alesapin 

Part of #52608

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
This is an automated comment for commit 6d8bf805d4d0f65119c1b38588be6b03d4706a5b with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/52792/6d8bf805d4d0f65119c1b38588be6b03d4706a5b/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Common/Allocator.h [hunk 1]
```diff
@@ -182,6 +189,30 @@ class Allocator
...
+        if (len_ < POPULATE_THRESHOLD)
+            return;
+
+        static const size_t page_size = ::getPageSize();
+        if (len_ < page_size) /// Rounded address should be still within [buf, buf + len).
+            return;
+
+        auto [buf, len] = adjustToPageSize(buf_, len_, page_size);
+        if (auto res = ::madvise(buf, len, MADV_POPULATE_WRITE); res < 0)
+            LOG_TRACE(&Poco::Logger::get("Allocator"), "Attempt to populate pages failed: {} (EINVAL is expected for kernels < 5.14)", errnoToString(res));
```
Comment (alexey-milovidov) [inline src/Common/Allocator.h:213]: The rate of these messages should be limited.

**Comment by canhld94** [comment]
@nickitat build issue because of line `Allocator.h:87`.
Looking for this to be merge soon :).

**Review by alexey-milovidov — APPROVED** [review]

**Comment by azat** [comment]
I like the idea of removing two allocations path!

But there were some noticeable performance decline, I saw you comment https://github.com/ClickHouse/ClickHouse/issues/52608#issuecomment-1657259343

But maybe the root cause is lack of the `mremap` now? At least I have this thought when I tried to do the same - https://github.com/ClickHouse/ClickHouse/pull/46457#issuecomment-1515347803 (and that was actually the reason why I closed my PR)

**Comment by nickitat** [comment]
> I like the idea of removing two allocations path!
> 
> But there were some noticeable performance decline, I saw you comment [#52608 (comment)](https://github.com/ClickHouse/ClickHouse/issues/52608#issuecomment-1657259343)
> 
> But maybe the root cause is lack of the `mremap` now? At least I have this thought when I tried to do the same - [#46457 (comment)](https://github.com/ClickHouse/ClickHouse/pull/46457#issuecomment-1515347803) (and that was actually the reason why I closed my PR)

as I understood the degradation happened because when we previously called `mmap` manually we were able to pass `MAP_POPULATE`. and after deleting these manual calls we got some deteriorated aggregation queries (only hash table allocator had `mmap_populate=1`). now we again do populate manually, but with the help of `madvise`.

**Comment by azat** [comment]
Hm, but `MADV_POPULATE_WRITE` (for `mmap_populate=1`) is part of this PR, but there are still performance degradation in it, so there should be something else

**Comment by compasses** [comment]
MADV_POPULATE_WRITE requires a newer version of the kernal? And I have tried using MADV_HUGEPAGE which seems work.

**Comment by nickitat** [comment]
> MADV_POPULATE_WRITE requires a newer version of the kernal?

yes, e.g. Ubuntu 22.04 LTS has it but 20.04 - don't, AFAIU

> And I have tried using MADV_HUGEPAGE which seems work. Improving performance a little, especially the TLB miss aspect.

yep, it is expected: https://github.com/ClickHouse/ClickHouse/pull/52157#issuecomment-1722526770

**Pull request merged**